### ISSUE TITLE
Storage Backend: Amazon Cloud Drive

### ISSUE THREAD

**Issue by e2b** [restic/restic#212]
The "Unlimited Everything" plan of [Amazon Cloud Drive](https://www.amazon.com/clouddrive/home/) is a quite affordable backup storage option. Amazon Cloud Drive has its own RESTful API.

**Comment by klauspost** [comment]
Just thought the same thing.

I may look at this once #21 is in place - uploading without compression seems like a waste of bandwidth.

**Comment by fd0** [comment]
That depends on your use case ;)

**Comment by stapelberg** [comment]
There’s some proof of concept code I found: http://sprunge.us/fdQF — it requires that an oauth token is in /tmp/token.json, but seems to work for me.

Motivated people could turn that into a clean backend for ACD :).

**Comment by jsimonetti** [comment]
+1 from me :)
I compile the backend @stapelberg found, and it indeed appears to work and use code from rclone.

**Comment by kisscool** [comment]
+1 too
I'm trying to use Restic + acd_cli (FUSE python client for ACD), but it's very unreliable for now : some operations do not behave as expected (file truncate, rename) and Restic randomly panics as a result.
A working ACD storage backend would do wonder.

**Comment by fd0** [comment]
I'm currently reworking the interface to the backends, this includes a radical simplification. This is basically done, but not yet merged. For the plan, see #383, the PR is #395.

Afterwards it will be much easier to implement new backends.

Before implementing many new backends, I'd like to have a list of rules that services we write backends for must fulfill, this may include that a test instance of the service must be available that we can run the integration tests against.

Do you by chance know whether there is a test service for ACD we can use for tests?

**Comment by kisscool** [comment]
As far as I know, there is no test instance for ACD. No mention of such a thing here : https://developer.amazon.com/public/apis/experience/cloud-drive/content/restful-api

But the https://github.com/ncw/rclone project already did an ACD backend in go. It seems to be fairly reusable as demonstrated by the proof of concept shared by @stapelberg .

**Comment by klauspost** [comment]
Actually, looking at the revised interface, it would be reasonably easy to do a full wrapper for rclone filesystems. Maybe that way separate implementations isn't needed?

**Comment by kisscool** [comment]
I don't know what @fd0 vision for Restic future is, but it would seem logical to focus the project on the backup intelligence instead of re-implementing a ton of remote filesystems one by one. Besides both project licenses are compatible. 
It would also solve the worries about how to test all those backends.

@klauspost was your idea to create a wrapper around rclone/fs/fs.go ? Is it doable without being tightly coupled with the internal logic of rclone ?

**Comment by klauspost** [comment]
Each backend implements the [fs.Fs](https://godoc.org/github.com/ncw/rclone/fs#Fs) interface.

It should be fairly easy to create a restic backend that uses an rclone filessystem+folder, provided it is already set up in the rclone configuration.

**Comment by fd0** [comment]
Hm, interesting idea, I have to think about it. Not having to implement all the backends by ourselves looks like a good idea, on the other side (at least at the moment) I must admit that I don't like the thought of a tight coupling between restic and rcclone, as this introduces a dependency that we can't control...

**Comment by stv0g** [comment]
I pledged a [5$ bounty](https://www.bountysource.com/issues/23684796-storage-backend-amazon-cloud-drive) for this feature.

Some thoughts:

- Amazon Cloud Drive is using AWS S3 / CloudFront as its backend. The `GET` requests are always redirected with a `Location` header to Cloudfront. So you could use the `Range` header to request only a portion of a pack file as it is required by the new backend API.
- You could start with integrating [go-acd](https://github.com/ncw/go-acd). That's the library which is used by rclone.

**Comment by romusz** [comment]
in case the priority of this FR depends on the popular vote: +1

**Comment by heikobornholdt** [comment]
👍

**Comment by Intensity** [comment]
👍

**Comment by nunofgs** [comment]
Yes please 👍

**Comment by stv0g** [comment]
How about adding some more bounties to this feature?

See: https://www.bountysource.com/issues/23684796-storage-backend-amazon-cloud-drive

**Comment by stapelberg** [comment]
The bounty is now 35 USD: https://www.bountysource.com/issues/23684796-storage-backend-amazon-cloud-drive

**Comment by fd0** [comment]
Hey, thanks for your interest in restic in general and this backend in particular.

Just to give you a heads-up what's the blocker here: I'm not sure how to handle third-party web services. For the `local` and `sftp` backends we have extensive tests in place that are ran for every push/PR as part of the CI tests on Travis. This is also true for the `s3` backend, there we're using a local [Minio](https://minio.io/) s3 server instance, due to this I've found several bugs in the minio client library we're using in the `s3` backend.

How can we run CI tests for backend implementations that require a third-party service? Is there e.g. a test service for ACD we could use? Or maybe just take well-tested code from other projects such as [rclone](https://github.com/ncw/rclone)?

**Comment by stapelberg** [comment]
One solution might be to register an account with Amazon, whitelist it for the Cloud Drive API and then use that for the CI tests? The downside is that such a test depends on Cloud Drive being available, but I guess we can wait for an hour or so occasionally before merging a PR? :)

**Comment by fd0** [comment]
That's the only solution I can imaging right now that allows us to run the tests against a live service (and that's desirable in my opinion).

When we add more backends for other services the following will happen:
 * The number of dependent services for running the CI tests grows
 * For each backend we will need a test account to use in the tests

Did I forget anything?

**Comment by stapelberg** [comment]
Your list looks good. There are of course more effects, but I’m not sure whether they are in scope for the question you’re trying to answer:

* More backends means changes that touch the backends API become more involved (need to update more code, test more code).
* People who want to run the tests locally need to create test accounts or use their own account.
* More backends make restic more appealing to more people :).

> This test account must allow parallel connections, as the tests (e.g. for a PR) are run in parallel

I think a simple way to take care of this requirement is to use different directories for each test invocation. Sending requests in parallel is usually not an issue with these services, and the different directories make sure the tests don’t clash.

**Comment by fd0** [comment]
> I think a simple way to take care of this requirement is to use different directories for each test invocation. Sending requests in parallel is usually not an issue with these services, and the different directories make sure the tests don’t clash.

What I meant was more of a question how many parallel connections a service accepts. For most web-based services this won't be limited (at least concerning the number of connections we require), but this may not be the case for other, more obscure services.

**Comment by stapelberg** [comment]
When a service limits the number of connections so aggressively that our testing is impacted, we could ask the service owner for an exception or rate-limit on our end as well. As a last resort, we could disable the tests for the backend in question or remove that backend altogether.

But, I suggest we cross that bridge when we get there :).

**Comment by Siviuze** [comment]
I tried to use ACD with restic (through Fuse) and the system is still reliable (same errors as kisscool). I tried rClone to check their backend and there isn't any problem.

But I do not like rClone (snapshots...).

Conclusion: +1 for a native ACD backend!

**Comment by wdl1908** [comment]
+1 for native ACD backend because the acd fuse driver does not seem to work at all with restic.

acd mount under /mnt/restic

restic 0.3.0 (v0.3.0-29-g190673b)
compiled at 2016-11-15 21:58:36 with go1.7.3 on linux/386

restic init gives me this error.

create key in backend at /mnt/restic failed: Calibrate: scrypt.Calibrate: scrypt: the parameters provided are invalid

**Comment by Twister915** [comment]
Is there a problem with trying to write this? The issue seems to imply it's basically done, but delayed due to other architecture goals. This was a while ago, though, so... should I write this backend from scratch and pull request, or is it still being done internally?

**Comment by fd0** [comment]
Writing the backend is the not so hard part, figuring out a good user interface for configuring access to the service is what's still missing here. What's the workflow for acquiring an oauth token for ACD?

**Comment by Twister915** [comment]
rclone does this using a local webserver, and the backend for duplicity provided a link to a solution hosted by the developer. The former is cumbersome for headless, and the latter makes this into a service that you'd want to continue to provide for free, although the resources would cost quite a bit. So... I'm not sure. That makes the former the only viable option.

Are you opposed to a webserver, in the style of rclone, to accomplish this? Perhaps restic could ask you what interface to bind on, or give you the opportunity to use a different copy of the same program (this is what rclone does) on a non-headless machine.

I'll have to check how rclone automates this process for testing, if it does that at all.

**Comment by fd0** [comment]
Yeah, testing is another story. How do we run the CI tests for these backends? Please don't get me wrong, I'd love to add cloud-based backends, but we need a clear strategy for this.

For configuration, I can also imagine having a CLI-based process, where restic prints instructions to the user.

What do you think?

**Comment by Twister915** [comment]
###Idea #1

What about simply integrating with rclone? Kinda a "unix philosophy" type of idea, you continue to be great at doing backups, they continue to be great at cloud copy, and users get best of both worlds. 

I'd have to see what their API affords, if it exists, but since we'd be leaving configuration solely to rclone (users would use rclone to add their cloud accounts, then would configure restic to "use rclone") you'd add far less complexity to restic.

###Idea #2 
I'll preface this by saying I've not read the docs for ACD, but I have implemented some basic oauth stuff in the past. 

If we implemented a webserver, we could simply test the following thing:
- Does the webserver respond on the right ports/addresses
- When I GET the / route, does it contain what I'd expect it to (a redirect, the text we put there, etc)
- If I mock the response that would come from amazon (more thought needed on how to do this correctly) how does it behave?
- Still needs a lot more thought: static "testing" credentials of some sort, to test the API. I'll have to look at how you test other services first.

and a few others, but the point is, we mock what we have to, and test the rest. I don't know if that'd test _everything_ but it'd test as much as we possibly could. I've not written extensive tests for my projects in the past, so please correct me if this isn't testing the right things haha

**Comment by Twister915** [comment]
Sorry for glossing over your reply, was typing mine a bit before I saw yours!

>For configuration, I can also imagine having a CLI-based process, where restic prints instructions to the user.

I imagine the process would probably be something like this, and this... again, is just rclone. I only reference it over and over because it's the only program that does _exactly_ what we're talking about quite well.

To start:
![Using config](https://tmg.pw/dN6NDwHeR)

Then, setup ACD: 
![ACD](https://tmg.pw/J0RPw4g0c)

Using auto config (non-headless)

It, very quickly, opened a local webserver in my browser which immediately redirected to Amazon for login, and once I signed in, it communicated with the CLI app producing:

This, in my browser:
![browser](https://tmg.pw/ZONhB3BYY)

and this, in my command line
(60% of the credential is missing, I don't think this is a security risk, haha)
![code](https://tmg.pw/aytaFd37F)

Then I save it, and it's now usable in the program.

**Comment by fd0** [comment]
Thanks for describing the process in such great detail, that is already very similar to what I had in mind.

I'm wondering: Why is the webserver needed at all? This process works for a "workstation" type of machine, but not on a server (where there is not browser). The workflow used by rclone is described here: http://rclone.org/remote_setup/

I don't know why we need a webserver for this, but I haven't implemented an oauth-based login workflow yet.

We'll also need a config file to store the token configured for the remote in, that's also not yet done.

**Comment by Twister915** [comment]
From my understanding, which is limited, the oauth data is provided to the user using the GET data in a redirect. 
Have a look at the URL in my screenshot of the browser. That was put there by Amazon. After I hit sign-in on my amazon cloud drive, it redirected me, immediately, to that 127.0.0.1 URL.

Perhaps, and this is likely since they (rclone) implemented a webserver instead of picking another simpler solution, that is the only way to get this data. When I implemented oauth before, this seemed to be the implication. If I am correct, then it follows that you must run your own webserver to provide a page to redirect to amazon, and a page to handle the redirect from amazon to do this.

**Comment by fd0** [comment]
That sounds plausible. Let me think about a strategy here, this may take some time.

We'll need to:
 * have a config file to store the authentication tokens
 * have "instances" of backends, e.g. something called `my_amazon_account` which is a ACD backend configured with a login token, so users can run `restic --repo my_amazon_account:/foo/bar/dir ...`
 * have a workflow to create these login tokens

Anything else I'm missing here?

**Comment by Twister915** [comment]
I think you got the big stuff outlined there. 

Would you want to move all current backends into a single abstraction that supports this, or would this whole system become a "cloud" backend in the current sense of a backend (which itself is configured through special restic commands)?

**Comment by jsimonetti** [comment]
* Have a workflow to refresh auth tokens (if they expire and the provider supplies a refresh token, which should be stored next to the auth token in your point 1)

**Comment by fd0** [comment]
Here's some background in regards to embedding a client secret in open source applications: http://stackoverflow.com/a/28109307

**Comment by Twister915** [comment]
Go std library has this: https://godoc.org/github.com/ncw/rclone/fs#MustObscure

and it's what rclone uses: https://github.com/ncw/rclone/blob/f3365dd2516709c8eb6d9b55496344b853ec89c7/googlecloudstorage/googlecloudstorage.go#L53

**Comment by klauspost** [comment]
There is no real solution, it is a broken concept to assume that any client can keep a secret.

However, if you consider what the client secret contains, it is not that important. The only real thing it allows is for Amazon (and others) to be able to identify a specific client, nothing more. It does not grant any special access - your tokens are used for that.

Sure a publicly available "client secret" can make other application identify themselves as restic, but other than risk that "restic" will be banned (or more likely rate limited) as a client, there is not much risk at exposing the client "secret". It will never put any user data in jeopardy.

**Comment by fd0** [comment]
The problem here is that somebody needs to register the clientID, for example me. If I'm using my normal Amazon account (just to name an example service), and "violate" the TOS for the service by publishing the client secret, they can terminate my account. That's not something I'm going to risk.

Another problem is that once the client secret changes (or is revoked), we're stuck with older versions of restic e.g. in Debian stable which are unable to communicate with the service. This is the case even if access to the service is restored shortly after, but the client secret is changed.

I've thought about possible solutions and found only two:
 * Live with the risk and just put the client secret into the source
 * Build restic in a way that users need to register their own client ID and client secret, via a nice UI that minimizes the hassle

Currently, I'm in favor of the second option, we need a UI for the oauth token thing anyway. What do you think?

**Comment by klauspost** [comment]
> If I'm using my normal Amazon account [...] 

I know that Nick has had some correspondence with Amazon, since rclone was being rate limited due to many users. It is however my impression (from memory) that they were quite forthcoming and encouraged OS development, and have made exemptions for his client. So I guess my advise would be to contact them and see how things go from there. In the overall picture I don't think they would mind the business coming from restic users.

**Comment by fd0** [comment]
Interesting idea, do you have any hint on who to contact at Amazon?

For Microsoft OneDrive he said that he did not contact anyone: https://github.com/ncw/rclone/issues/372

**Comment by stapelberg** [comment]
I know that @breunigs had bad luck with his amazon cloud drive duplicity backend — they wouldn’t give him any rate limit exemptions AFAIK.

**Comment by breunigs** [comment]
I have only read the last few comments, so please forgive me if this info is not needed:
* rclone implements the web server on top of it offering remote setup where you copy the URL. Having a local webserver is just more convenient
* if you want to whitelist any redirect target in Amazon, it has to be on a http*s* machine – linking to http URLs is not okay. Only exception is localhost. So, for remote setup you can either redirect the user to a blank page and hope they realize what they have to do, or host some page with instructions. I added https://breunig.xyz/duplicity/copy.html for duplicity, since it doesn't have https infrastructure yet. Amazon will add all details in the query string, so you can get away with making this a static page
* You need an Amazon *Developer* account. You can use your existing credentials to log in I believe, but you can also create a new one
* There is a process where you register your app and then at some later stage you can create a security profile for said app. This process is very confusing, because of horrible UX, but it should work without human interaction from Amazon.
* What the limits are is unclear, Amazon don't say. It's clear there are multiple stages: per user, per API endpoint, per credentials
* If you want production limits, you send an email with "details" to clouddrive-api-invite@amazon.com. Use a big player mail server, or they will tag you as spam and it takes a month or two until some poor soul went through all their spam.

Also, a final word of advice: read through rclone's workarounds for Amazon Drive. The API contains a lot of undocumented "eventual consistency" gotchas. It even goes out of its way to cache an outdated response it gave you, so that you need to wait even longer if you were too hasty to begin with. This is on top of it reporting errors when there are none, one just needs to wait. 

HTH,
Stefan

**Comment by fd0** [comment]
Thanks for the information!

**Comment by jsimonetti** [comment]
Just throwing something out there:

What is we remove all (but local and REST) backends from restic and stick them into restic/rest-server?

This allows restic to focus on doing backups properly and filesystem implementations are done in the rest-server.
This also leaves restic with just 1 backends API to maintain.

This doesn't solve the testing problem, but will certainly help keep the restic source clean/focussed and it is easier to make API changes inside restic.

**Comment by fd0** [comment]
Thanks for the suggestion. Unfortunately I don't like it at all, in my opinion this approach will lead to even more errors. The backend API interface was stable for a long time, then changed recently, and will be stable again. The interface is already rather small.

We should try to get backends into restic (including proper CI tests) as soon as possible, that's IMHO the only way to make sure they work.

In case of the Amazon ACD backend, we need to answer the outstanding questions first.

**Comment by fd0** [comment]
The Amazon Developer Guide for Amazon Drive (what's it called these days) states that:

> What Not To Build
>
> We understand that developers spend a lot of time creating apps that customers love to use. We want to give you some guidelines for apps not to build on the Amazon Drive platform to save you time in building something that will not be approved during the App Review process. While this isn’t an all-inclusive list, focus your energy on developing apps that solve customer problems, which don’t violate our Terms of Use, Brand Guidelines and the App Distribution and Services Agreement.
> [...]
> * **Don’t build apps that encrypt customer data**

I feel that Amazon Drive is not the right platform for storing encrypted backups.

**Comment by askielboe** [comment]
Interesting. This must be a new addition, as it definitely was not the case when ACD support was added to Arq.

Seems ACD is not a real storage option after all.

**Comment by e2b** [comment]
Indeed an addition within the last year, wasn't listed one year ago: http://web.archive.org/web/20160322034250/https://developer.amazon.com/public/apis/experience/cloud-drive/content/developer-guide

**Comment by stapelberg** [comment]
Amazon has since clarified this in https://forums.developer.amazon.com/questions/54909/impact-of-dont-encrypt-customer-data-part-of-drive.html:

> > What if the customer choses to encrypt their data?
> They can do that, and that is fine.

So, restic and other apps should be good.

**Comment by stv0g** [comment]
I think their intention is to protect the users having their data encrypted without a way to recover it.

 

Steffen

**Comment by stapelberg** [comment]
One other motivation which I find plausible is to increase interoperability — if each application encrypts their files, the user’s ability to switch between applications is severely hampered.

**Comment by askielboe** [comment]
I asked Arq Backup support. They encrypt everything, and said that their app had been approved by Amazon, and to not worry.

I'm not sure what Amazon is trying to say. But seems that are now evaluating each case as they come in.

**Comment by mikesager** [comment]
Not sure if anybody is aware of the recent ACD drama with acd_cli and rclone, but a TL;DR of the situation is that they have had their ACD API access revoked due to TOS violations.  Their efforts to regain API access are apparently being hampered by the fact that Amazon has stopped accepting new third-party apps for ACD.  I assume this latter revelation stops any Restic ACD support in its tracks, unless the project had already obtained ACD API access.

**Comment by sedlund** [comment]
acd_cli API access was revoked due to a [security issue](https://github.com/yadayada/acd_cli/issues/549) with their ouath app, not a TOS violation.  The problem has been [fixed](https://github.com/yadayada/acd_api_oauth/commit/339c4b3ad3e6960e1e1438ac8e015aad0d242f50) and Amazon re-instated their key.  Although this is off topic from this project.

New ACD API access is currently [closed](https://forums.developer.amazon.com/answers/62343/view.html).

**Comment by fd0** [comment]
Thanks for posting this here, I wasn't aware of it. I had reservations implementing ACD, and it seems that Amazon indeed did not like secrets in the code of an Open Source program: rclone was banned for it: https://forum.rclone.org/t/rclone-has-been-banned-from-amazon-drive/2314

On the other hand, acd_cli implemented an OAUTH auth service (not sure what the correct nomenclature here is). This handles authorization for all users, and there apparently was a bug that allowed people to access/modify other people's files.

Since Amazon isn't accepting new clients anyway I'm closing this issue for now. Thanks!
